feat: use babel to extract translation sources

ankush opened the pull request:
This piece of beauty has outlived its maintainability, I spent **hours** to extend it to accommodate new lines. Time to use something from a civilized age. 
![image](https://user-images.githubusercontent.com/9079960/182211655-3c3298cc-40a6-43c5-a282-b234fde3b1fa.png)



- [x] Use babel's parser for python
- [x] Use babel's lexer + modified parser for JS code
- [x] JS calls inside of template strings (HTML)
- [X] keep using regex for all other sources 


Why do this? 

- Babel operates on tokens instead of lines so it can handle newlines very well which regex can't. 
- Babel is already a dependency
- Babel is extensible if we end up adding a new translation function (e.g. lazy translations) then we need to change few lines to make it work with current extractors)


Pros:
- Far far more accurate parsing, using the same parser as the language itself.


Cons:
- Significantly slower than regex, but this doesn't happen that often so eh. 



`no-docs`


Concludes / continues: https://github.com/frappe/frappe/pull/16409 

---

This PR also changes how client-side translations work and makes them similar to how server-side translations work. This increases boot size for non-English users but makes translation more accurate and reliable. We will figure out a better way to cache boot messages on the client side in some future improvement. 

closes https://github.com/frappe/frappe/issues/17684

ankush committed:
feat: use AST to extract translation sources

ankush commented:
After / before (python change)

<img width="734" alt="Screenshot 2022-08-01 at 11 27 21 PM" src="https://user-images.githubusercontent.com/9079960/182212004-1e052f69-222d-416a-9c33-c12afe3e87c5.png">


Some new discovery 👀 

<img width="1321" alt="Screenshot 2022-08-01 at 11 31 46 PM" src="https://user-images.githubusercontent.com/9079960/182212704-327ae859-e63b-4b61-9145-76ace0714ad0.png">



And some dropped garbage that was incorrectly matching old regex. 

<img width="672" alt="Screenshot 2022-08-01 at 11 32 56 PM" src="https://user-images.githubusercontent.com/9079960/182212852-e06a2f15-2f7b-4d1d-b98c-1a1b0dc819e5.png">

sagarvora commented:
Worth mentioning (I've had good success with these utils):
https://babel.pocoo.org/en/latest/api/messages/extract.html#babel.messages.extract.extract

![image](https://user-images.githubusercontent.com/16315650/182284636-8ef98d41-7fcf-4612-9263-47c730a50e11.png)

They also have a JS lexer:
https://github.com/python-babel/babel/blob/40059883cc260af890bcc81a1a21c0f84781857f/babel/messages/jslexer.py#L1

ankush commented:
@sagarvora looking into it. This will be better solution and babel is already a dependency. 


One problem I know of is jinja, we might have to continue to use old regex for files containing jinja I think.

sagarvora commented:
> One problem I know of is jinja, we might have to continue to use old regex for files containing jinja I think.

Or we can extract python code from Jinja. Should be simple.
Might help: https://jinja.palletsprojects.com/en/3.1.x/api/#low-level-api

We can't for JS templates though. And Vue files (?)

ankush commented:
JS with jinja is what I was talking about, I think we can continue to use old regex for them. VueJS is similar problem. 


btw I had much better consistent success with [treesitter](https://github.com/tree-sitter/py-tree-sitter ) queries across all languages but this will add system dependency 🥲

sagarvora commented:
> btw I had much better consistent success with [treesitter](https://github.com/tree-sitter/py-tree-sitter) queries across all languages but this will add system dependency

but it looks so good 😍 
<s>maybe we can keep it optional - people who want to translate can install</s> I think extraction happens everywhere 😅

ankush committed:
refactor: extract python translations using babel

RIP my cool handwritten AST code :'(

Few things to note:

1. Publicly documented APIs, they don't support capturing kwargs.
2. We can't use documented "lower level" APIs, we need to go _even lower_.

ankush commented:
Rewrote it using babel. 

One important things to note:

We can't use documented "lower level" APIs, we need to go _even lower_. Publicly documented APIs don't support capturing kwargs. 😄 I added a little bit of wiring on top of individual extractors to make it work with our syntax.  This is still easy to maintain ig.

ankush commented:
The problem with JS extractor is there's no way to extract 2nd and 3rd argument without flattening whole thing 🥲 


<img width="1326" alt="Screenshot 2022-08-02 at 6 03 33 PM" src="https://user-images.githubusercontent.com/9079960/182376552-90dbd398-09e0-4e7b-84eb-eef84c1a6ba3.png">



Might have to write it ourselves using babel's JS tokenizer.

ankush committed:
feat: JS extractor and custom JS parser

ankush committed:
feat: modified parser for frappe JS translate syntax

ankush committed:
fix: ignore built assets in translation

ankush commented:
Errr, babel doesn't handle expressions inside the template strings like these

```javascript
`Template string with args ${__('that contain translation')}`
```

Might have to implement it ourselves.

Edit: reused old regex for these missing cases there's also:

- translation added in comments  (intentionally 🙄 )
- jinja calls inside js as strings

ankush committed:
refactor: simplify get_messages_from_file

ankush committed:
fix: send all messages on boot instead of scanning

close https://github.com/frappe/frappe/issues/17684

ankush committed:
chore: remove pointless template strings

sagarvora commented on frappe/translate.py:
```diff
@@ -679,41 +692,241 @@ def get_all_messages_from_js_files(app_name=None):
+	return messages
+
+
+def extract_javascript(code, keywords=("__"), options=None):
```
tuples be like that 😅 

```suggestion
def extract_javascript(code, keywords=("__",), options=None):
```

ankush committed:
chore: correct default arg for extract_javascript

Tuples need extra comma to make sure it's not treated as string. 

[skip ci]

Co-authored-by: Sagar Vora <sagar@resilient.tech>

barredterra commented on frappe/tests/test_translate.py:
```diff
@@ -125,6 +133,91 @@ def test_load_all_translate_files(self):
+					"format", "replacements"
+				],
+			)
+			_(`template strings not supported yet`)
```
Second underscore missing here

ankush replied:
Ah, the message inside test string is misleading, it's meant to test that `_` doesn't get captured. Previous regex didn't respect the actual translation function for the language type. 

Equivalent python version:
https://github.com/frappe/frappe/blob/f6210e7c38d33f99f54d22067228b0f9ec558d4d/frappe/tests/test_translate.py#L146



Template strings still won't work if they have replacement inside them, else they will work.... similar to python
